### DISCUSSION TITLE
Wavelength to XYZ - what CMFs should I actually be using?

### DISCUSSION THREAD

**Discussion by aarondemolder** [colour-science/colour#767] [Q&A]
Hey!

This is an excellent resource, and wanted to first of all thank you for the work and making it available. This has been quite a learning experience.

I'm having a bit of confusion with the wavelength_to_XYZ function with various datasets and hoped you could point me in the right direction. 

If I have 3 laser sources, 643nm, 520nm and 450nm, and I want to display sRGB content on them, I want to create a matrix that will convert sRGB -> XYZ -> Display.

So I use a cmfs to get my XYZ values for the primaries.

```
cmfs = MSDS_CMFS['CIE 1931 2 Degree Standard Observer']

R = wavelength_to_XYZ(643, cmfs)
G = wavelength_to_XYZ(520, cmfs)
B = wavelength_to_XYZ(450, cmfs)
```

I can then go through the process to get the output matrix, but I notice that there's significant differences to my blue using various data sources because of the short wavelength differences. Where I am getting purple instead of blue. Does the data here take into account the corrections to V(λ) given by http://cvrl.ioo.ucl.ac.uk/database/text/cmfs/ciexyz31.htm or does it only use the original dataset?

I chose 1931 2 deg because it's what I think we can assume for sRGB, unless that assumption is incorrect? Should I instead be using CIE 2012 2 Degree Standard Observer? 

Bonus problem: 

How can I best avoid hue shift from blue? Where my primary is not along the same axis as for example a rec709 -> BT2020 conversion. Best shown by this section of comparison on a 1976 u,v plot:

(white triangle is my display gamut, yellow= ACEScg, blue/red dot is DCI-P3 and rec709)
![image](https://user-images.githubusercontent.com/25490210/106510034-4c386a80-64c6-11eb-91c7-cca112564db1.png)

Any guidance on this would be absolutely fantastic.

Thank you!
Aaron

Edit: made a silly suggesting sRGB had wavelength primaries and wasn't just a fixed matrix

**Comment by thomasmansencal** [comment]
Hi @aarondemolder and Welcome!

First, thanks for the kind words :)

> If I have 3 laser sources, 643nm, 520nm and 450nm, and I want to display sRGB content on them, I want to create a matrix that will convert sRGB -> XYZ -> Display.

As you might have discovered, they will be very much out-of-gamut for sRGB!

> Does the data here take into account the corrections to V(λ) given by http://cvrl.ioo.ucl.ac.uk/database/text/cmfs/ciexyz31.htm

The data for the CIE 1931 2° Standard Observer is that of the CIE and thus does not have the Judd-Vos correction.

> I chose 1931 2 deg because it's what I think we can assume for sRGB, unless that assumption is incorrect?

This is correct and raises an important question, one that annoys the CIE extremely: many RGB colourspaces are not spectral defined thus they cannot be migrated under another observer trivially. What are the primaries of sRGB under the CIE 1964 10° Standard Observer?

> How can I best avoid hue shift from blue?

I guess it would depend on what you are trying to achieve, but you are fortunate enough that your display (might encompass) encompasses sRGB (haven't checked the other primaries), so that display should be able to represent all the sRGB gamut without trouble.

**Reply by aarondemolder** [reply]
Thanks for getting back to me so quickly!

The lack of Judd-Vos correction *might* explain my issue then, with my blue wavelength being rather short. But obviously then I'll have to cross my fingers very tightly that the differences between using slightly different observers won't have detrimental effects.

I've actually tried the CIE 1964 10° Standard Observer to see what happened and I ended up with quite a green hue shift going from sRGB, which I think is to be expected, because of course the non-spectral primaries haven't been adjusted for the different CMF.

Luckily yes - my colourspace does encompass sRGB, but is very very close on the blue primary to the edge.

I'll make a couple of variations and see if I can get what I'm looking for - thank you for the clarification!

**Reply by thomasmansencal** [reply]
You are welcome!

> I ended up with quite a green hue shift going from sRGB

How do you define sRGB though under the CIE 1964 10° Standard Observer? Some of us, e.g. @KevinJW, talked about it quite a few times by for example cherry-picking a bunch of phosphors and fitting them to the current primaries, but we never really committed to it.

**Reply by aarondemolder** [reply]
So after chatting to someone here I realised I'd made an error which resulted in my green shifting (and the blue problem!), and I can now transform sRGB content to view it without visible change on one of our laser displays.

But what I'm doing is actually skipping the 10 degree definition of sRGB, and moving the redefinition to my laserspace, as I have known XYZ positions for my primaries which are monochromatic wavelengths, and I obviously know these for any CMF - rather than having to worry about phosphors at all. Maybe that's cheating!

sRGB -> "Laserspace" 1931 2° -> "Laserspace" Any CMF

I've viewed sRGB content transformed to CIE 2012 10° this afternoon this way and it identical, as it should.

Ultimately though my ideal workflow will be just have all content in ACES and go from there, but this is also defined around CIE 1931 2° so I have to do the same redefinition step in my ODT if I want my output CMF to be different.

I'm yet to see any benefit of this yet having only tried from sRGB though!
